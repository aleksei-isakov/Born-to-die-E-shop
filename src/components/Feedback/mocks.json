{
    "feedbacks" : [
        {
            "reviewerName" : "Vitali",
            "date" : "March 5, 2017 10:00:00",
            "rating" : 2,
            "comment" : "Lorem ipsum",
            "reviewerId"  :"64defdaf-73f8-4043-b1b7-2e7317bc1bc8",
            "id" : "a8ac19f4-ee6d-41bd-8700-6f1efc7693a5"
        },
        {
            "reviewerName" : "Damian",
            "date" : "March 22, 2022 10:00:00",
            "rating" : 4,
            "comment" : "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "reviewerId"  : "64defdaf-73f8-4043-b1b7-2e7317bc1bc7",
            "id" : "a8ac19f4-ee6d-41bd-8700-6f1efc7693a6"
        },
        {
            "reviewerName" : "Anonymus",
            "date" : "March 6, 2010 10:00:00",
            "rating" : 5,
            "comment" : "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident.",
            "reviewerId"  :"64defdaf-73f8-4043-b1b7-2e7317bc1bc9",
            "id" : "a8ac19f4-ee6d-41bd-8700-6f1efc7693a7"
        }
    ]
}