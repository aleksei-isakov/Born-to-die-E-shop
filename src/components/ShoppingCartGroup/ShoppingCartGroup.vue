<template>
  <div class="group">
    <div class="group__title">
      <input v-model="isGroupChecked" type="checkbox" class="group__checkbox" />
      <h2>{{ seller.name }}</h2>
    </div>
    <div class="group__list">
      <shopping-cart-item
        v-for="cartItem in products"
        :id="cartItem.id"
        :key="cartItem.id"
        :name="cartItem.name"
        :quantity="cartItem.quantity"
        :price="cartItem.price"
        :discount-percentage="cartItem.discountPercentage"
        :price-with-discount="cartItem.priceWithDiscount"
        :images="cartItem.images"
        :is-checked="cartItem.checked"
      />
    </div>
  </div>
</template>

<script>
import ShoppingCartItem from '@/components/ShoppingCartItem/ShoppingCartItem';
import { mapActions } from 'vuex';

export default {
  name: 'ShoppingCartGroup',

  components: { ShoppingCartItem },

  props: {
    seller: {
      type: Object,
      default: () => {},
      required: true
    },

    products: {
      type: Array,
      default: () => [],
      required: true
    }
  },

  computed: {
    isGroupChecked: {
      get() {
        return this.products.every(({ checked }) => checked);
      },

      set(value) {
        this.products.forEach(({ checked, id }) => {
          if (checked !== value) {
            value ? this.checkCartItem(id) : this.uncheckCartItem(id);
          }
        });
      }
    }
  },

  methods: {
    ...mapActions('ShoppingCartModule', ['uncheckCartItem', 'checkCartItem'])
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.group {
  display: flex;
  flex-direction: column;
  gap: 20px;

  &__checkbox {
    width: 20px;
    height: 20px;
    cursor: pointer;
  }

  &__title {
    display: flex;
    align-items: center;
    gap: 20px;

    h2 {
      text-align: left;
      font-weight: 500;
    }
  }

  &__list {
    display: flex;
    justify-content: center;
    flex-direction: column;
    align-items: center;
    gap: 20px;
    padding-left: 30px;
  }
}
</style>
