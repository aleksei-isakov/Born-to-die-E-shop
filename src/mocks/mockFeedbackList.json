[
  {
    "id": "11111111",
    "reviewerName": "Vitali",
    "date": "March 5, 2017 10:00:00",
    "rating": 2,
    "comment": "Lorem ipsum",
    "reviewerId": "123"
  },
  {
    "id": "22222222",
    "reviewerName": "Damian",
    "date": "March 22, 2022 10:00:00",
    "rating": 4,
    "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
    "reviewerId": "456"
  },
  {
    "id": "33333333",
    "reviewerName": "Anonymus",
    "date": "March 6, 2010 10:00:00",
    "rating": 5,
    "comment": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident.",
    "reviewerId": "789"
  }
]
