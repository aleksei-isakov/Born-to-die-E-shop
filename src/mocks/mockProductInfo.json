{
  "id": "6f40cf0f-9841-4994-ae04-2203145592a5",
  "name": "Fantastic Soft Sausages",
  "description":
    "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
  "price": 852,
  "images": [
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480",
    "http://placeimg.com/640/480"
  ],
  "createdAt": "2022-01-31T09:22:07.577Z",
  "updatedAt": "",
  "feedbacks": []
}
