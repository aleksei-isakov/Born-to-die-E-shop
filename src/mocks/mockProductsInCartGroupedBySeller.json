[
  {
    "seller": {
      "id": "505c1bb1-8016-4078-b0e8-6dabd255c896",
      "name": "Brakus and Sons"
    },
    "products": [
      {
        "id": "59792c44-f81b-4b2a-ac48-77ba79eac3aa",
        "name": "Rustic Concrete Bacon",
        "category": {
          "id": "74751bdd-dada-44e0-bd0c-6e97eacd256d",
          "name": "molestiae",
          "description": "Vel dicta asperiores nihil voluptate nihil occaecati libero.",
          "createdAt": "2021-05-30T04:17:53.113Z",
          "updatedAt": ""
        },
        "seller": {
          "id": "505c1bb1-8016-4078-b0e8-6dabd255c896",
          "name": "Brakus and Sons"
        },
        "description": "The automobile layout consists of a front-engine design, with transaxle-type transmissions mounted at the rear of the engine and four wheel drive",
        "price": 959,
        "images": [
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480"
        ],
        "createdAt": "2021-05-22T07:23:59.367Z",
        "updatedAt": "",
        "feedbacks": [],
        "averageRating": 0,
        "quantity": 1,
        "checked": true
      },
      {
        "id": "6f33dad2-4e99-41f8-bb5c-fb0390572bd1",
        "name": "Sleek Metal Shirt",
        "category": {
          "id": "496fcf88-8c02-4175-8b80-3a325ee2f6a6",
          "name": "quo",
          "description": "Itaque neque laudantium minus sit fugiat.",
          "createdAt": "2021-12-13T18:17:58.737Z",
          "updatedAt": ""
        },
        "seller": {
          "id": "505c1bb1-8016-4078-b0e8-6dabd255c896",
          "name": "Brakus and Sons"
        },
        "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
        "price": 772,
        "images": [
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480"
        ],
        "createdAt": "2021-05-27T18:54:17.222Z",
        "updatedAt": "",
        "feedbacks": [],
        "averageRating": 0,
        "quantity": 1,
        "checked": true
      }
    ]
  },
  {
    "seller": {
      "id": "6f928573-3e0d-4056-aba4-a237819780b6",
      "name": "Collins Inc"
    },
    "products": [
      {
        "id": "da9222b2-18a1-4f4f-a16d-1f8e31313da7",
        "name": "Sleek Steel Keyboard",
        "category": {
          "id": "74751bdd-dada-44e0-bd0c-6e97eacd256d",
          "name": "molestiae",
          "description": "Vel dicta asperiores nihil voluptate nihil occaecati libero.",
          "createdAt": "2021-05-30T04:17:53.113Z",
          "updatedAt": ""
        },
        "seller": {
          "id": "6f928573-3e0d-4056-aba4-a237819780b6",
          "name": "Collins Inc"
        },
        "description": "The automobile layout consists of a front-engine design, with transaxle-type transmissions mounted at the rear of the engine and four wheel drive",
        "price": 928,
        "images": [
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480"
        ],
        "createdAt": "2021-05-02T02:06:03.911Z",
        "updatedAt": "",
        "feedbacks": [],
        "averageRating": 0,
        "quantity": 5,
        "checked": false
      },
      {
        "id": "1e7e4810-839b-4e0f-a3e3-6d3430b90d63",
        "name": "Incredible Granite Chair",
        "category": {
          "id": "27f8d06f-618b-40ad-baf5-4304e8ea61b4",
          "name": "aspernatur",
          "description": "Possimus qui qui impedit.",
          "createdAt": "2021-04-21T16:59:35.250Z",
          "updatedAt": ""
        },
        "seller": {
          "id": "6f928573-3e0d-4056-aba4-a237819780b6",
          "name": "Collins Inc"
        },
        "description": "The beautiful range of Apple Natural√© that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients",
        "price": 840,
        "images": [
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480"
        ],
        "createdAt": "2021-09-18T23:29:24.928Z",
        "updatedAt": "",
        "feedbacks": [],
        "averageRating": 0,
        "quantity": 4,
        "checked": false
      },
      {
        "id": "d5f85781-bccc-4e96-b56d-b5ce6cc58a8c",
        "name": "Rustic Metal Keyboard",
        "category": {
          "id": "e78d37f3-cd15-438f-9d64-b17148b22a9a",
          "name": "quia",
          "description": "Perferendis autem quo fugit provident consequatur dolorum et aliquid voluptas.",
          "createdAt": "2021-11-14T22:57:25.287Z",
          "updatedAt": ""
        },
        "seller": {
          "id": "6f928573-3e0d-4056-aba4-a237819780b6",
          "name": "Collins Inc"
        },
        "description": "New ABC 13 9370, 13.3, 5th Gen CoreA5-8250U, 8GB RAM, 256GB SSD, power UHD Graphics, OS 10 Home, OS Office A & J 2016",
        "price": 827,
        "images": [
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480",
          "http://placeimg.com/640/480"
        ],
        "createdAt": "2021-04-21T08:07:42.674Z",
        "updatedAt": "",
        "feedbacks": [],
        "averageRating": 0,
        "quantity": 1,
        "checked": true
      }
    ]
  }
]
