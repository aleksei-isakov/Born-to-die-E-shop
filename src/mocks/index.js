import mockCurrentUserInfo from './mockCurrentUserInfo';
import mockFeedbackList from './mockFeedbackList';
import mockProductInfo from './mockProductInfo';

export { mockCurrentUserInfo, mockFeedbackList, mockProductInfo };
