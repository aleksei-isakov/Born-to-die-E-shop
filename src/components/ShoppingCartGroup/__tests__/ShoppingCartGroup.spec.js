import { mount, createLocalVue } from '@vue/test-utils';
import ShoppingCartGroup from '@/components/ShoppingCartGroup/ShoppingCartGroup';
import { mockProductsInCartGroupedBySeller } from '@/mocks';
import Vuex from 'vuex';

const localVue = createLocalVue();

localVue.use(Vuex);

const groupWithAllCheckedItems = mockProductsInCartGroupedBySeller[0];
const groupWithPartlyCheckedItems = mockProductsInCartGroupedBySeller[1];

describe('ShoppingCartGroup', () => {
  let wrapper;
  let groupCheckbox;
  let actions;
  let store;

  beforeEach(() => {
    actions = {
      checkCartItem: jest.fn(),
      uncheckCartItem: jest.fn()
    };

    store = new Vuex.Store({
      modules: {
        ShoppingCartModule: {
          namespaced: true,
          actions
        }
      }
    });

    wrapper = mount(ShoppingCartGroup, {
      store,
      localVue,
      propsData: {
        seller: groupWithAllCheckedItems.seller,
        products: groupWithAllCheckedItems.products
      }
    });
    groupCheckbox = wrapper.find('.group__checkbox');
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  test('should be checked if all items inside it are checked', () => {
    expect(wrapper.find('.group__checkbox').element.checked).toBeTruthy();
  });

  test('should call action "uncheckCartItem" for all checked products in group after user unchecked group checkbox', async () => {
    await groupCheckbox.setChecked(false);
    expect(groupCheckbox.element.checked).toBeFalsy();

    const productsQuantity = groupWithAllCheckedItems.products.length;

    expect(actions.uncheckCartItem).toHaveBeenCalledTimes(productsQuantity);
  });

  test('match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('ShoppingCartGroup', () => {
  let wrapper;
  let groupCheckbox;
  let actions;
  let store;

  beforeEach(() => {
    actions = {
      checkCartItem: jest.fn(),
      uncheckCartItem: jest.fn()
    };

    store = new Vuex.Store({
      modules: {
        ShoppingCartModule: {
          namespaced: true,
          actions
        }
      }
    });

    wrapper = mount(ShoppingCartGroup, {
      store,
      localVue,
      propsData: {
        seller: groupWithPartlyCheckedItems.seller,
        products: groupWithPartlyCheckedItems.products
      }
    });
    groupCheckbox = wrapper.find('.group__checkbox');
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('should not be checked if not all items inside it are checked', () => {
    expect(wrapper.find('.group__checkbox').element.checked).toBeFalsy();
  });

  test('should call action "checkCartItem" for all unchecked products in group after user checked group checkbox', async () => {
    await groupCheckbox.setChecked(true);
    expect(groupCheckbox.element.checked).toBeTruthy();

    const productsQuantity = groupWithPartlyCheckedItems.products.filter(
      ({ checked }) => !checked
    ).length;

    expect(actions.checkCartItem).toHaveBeenCalledTimes(productsQuantity);
  });
});
