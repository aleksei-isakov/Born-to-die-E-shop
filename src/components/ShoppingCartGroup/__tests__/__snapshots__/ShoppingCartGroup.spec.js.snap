// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCartGroup match snapshot 1`] = `
<div class="group">
  <div class="group__title"><input type="checkbox" class="group__checkbox">
    <h2>Brakus and Sons</h2>
  </div>
  <div class="group__list">
    <div class="cart-item"><input type="checkbox" class="cart-item__checkbox">
      <div class="cart-item__wrapper"><img src="http://placeimg.com/640/480" class="cart-item__img">
        <div class="cart-item__name">Rustic Concrete Bacon</div>
        <div class="">

        </div>
        <div class="cart-item__price-container">
          <!---->
          <div class="cart-item__price">
            959
          </div>
        </div>
        <div class="quantity__wrapper"><button class="base-button base-text-button base-text-filled-button custom-blue">-</button> <input readonly="readonly" class="quantity__field"> <button class="base-button base-text-button base-text-filled-button custom-blue">+</button></div>
      </div>
      <div class="title-icon trash-icon cart-item__trash-icon"><img src="" width="25" alt="icon"></div>
    </div>
    <div class="cart-item"><input type="checkbox" class="cart-item__checkbox">
      <div class="cart-item__wrapper"><img src="http://placeimg.com/640/480" class="cart-item__img">
        <div class="cart-item__name">Sleek Metal Shirt</div>
        <div class="">

        </div>
        <div class="cart-item__price-container">
          <!---->
          <div class="cart-item__price">
            772
          </div>
        </div>
        <div class="quantity__wrapper"><button class="base-button base-text-button base-text-filled-button custom-blue">-</button> <input readonly="readonly" class="quantity__field"> <button class="base-button base-text-button base-text-filled-button custom-blue">+</button></div>
      </div>
      <div class="title-icon trash-icon cart-item__trash-icon"><img src="" width="25" alt="icon"></div>
    </div>
  </div>
</div>
`;
